codex/utils/safeasynciter.nim

```diff
@@ -0,0 +1,233 @@
+########################################################################
+
+proc new*[T](
+    _: type SafeAsyncIter[T],
+    genNext: SafeGenNext[?!T],
+    isFinished: IsFinished,
+    finishOnErr: bool = true,
+): SafeAsyncIter[T] =
+  ## Creates a new Iter using elements returned by supplier function `genNext`.
+  ## Iter is finished whenever `isFinished` returns true.
+  ##
+
+  var iter = SafeAsyncIter[T]()
+
+  proc next(): Future[?!T] {.async: (raises: [CancelledError]).} =
+    try:
+      if not iter.finished:
+        let item = await genNext()
+        if finishOnErr and err =? item.errorOption:
+          iter.finished = true
+          return failure(err)
+        if isFinished():
+          iter.finished = true
+        return item
+      else:
+        return failure("SafeAsyncIter is finished but next item was requested")
+    except CancelledError as err:
+      iter.finished = true
+      raise err
+
+  if isFinished():
+    iter.finished = true
+
+  iter.next = next
+  return iter
+
+# forward declaration
+proc mapAsync*[T, U](
+  iter: Iter[T], fn: SafeFunction[T, ?!U], finishOnErr: bool = true
+): SafeAsyncIter[U]
+
+proc new*[U, V: Ordinal](
+    _: type SafeAsyncIter[U], slice: HSlice[U, V], finishOnErr: bool = true
+): SafeAsyncIter[U] =
+  ## Creates new Iter from a slice
+  ##
+
+  let iter = Iter[U].new(slice)
+  mapAsync[U, U](
+    iter,
+    proc(i: U): Future[?!U] {.async: (raises: [CancelledError]).} =
+      success[U](i),
+  )
+
+proc new*[U, V, S: Ordinal](
+    _: type SafeAsyncIter[U], a: U, b: V, step: S = 1, finishOnErr: bool = true
+): SafeAsyncIter[U] =
+  ## Creates new Iter in range a..b with specified step (default 1)
+  ##
+
+  let iter = Iter[U].new(a, b, step)
+  mapAsync[U, U](
+    iter,
+    proc(i: U): Future[?!U] {.async: (raises: [CancelledError]).} =
+      U.success(i),
+    finishOnErr = finishOnErr,
+  )
+
+proc finish*[T](self: SafeAsyncIter[T]): void =
+  self.finished = true
+
+proc finished*[T](self: SafeAsyncIter[T]): bool =
+  self.finished
+
+iterator items*[T](self: SafeAsyncIter[T]): auto {.inline.} =
+  while not self.finished:
+    yield self.next()
+
+iterator pairs*[T](self: SafeAsyncIter[T]): auto {.inline.} =
+  var i = 0
+  while not self.finished:
+    yield (i, self.next())
+    inc(i)
+
+proc mapAsync*[T, U](
+    iter: Iter[T], fn: SafeFunction[T, ?!U], finishOnErr: bool = true
+): SafeAsyncIter[U] =
+  SafeAsyncIter[U].new(
+    genNext = () => fn(iter.next()),
+    isFinished = () => iter.finished(),
+    finishOnErr = finishOnErr,
+  )
+
+proc map*[T, U](
+    iter: SafeAsyncIter[T], fn: SafeFunction[?!T, ?!U], finishOnErr: bool = true
+): SafeAsyncIter[U] =
+  SafeAsyncIter[U].new(
+    genNext = () => iter.next().flatMap(fn),
+    isFinished = () => iter.finished,
+    finishOnErr = finishOnErr,
+  )
+
+proc mapFilter*[T, U](
+    iter: SafeAsyncIter[T],
+    mapPredicate: SafeFunction[?!T, Option[?!U]],
+    finishOnErr: bool = true,
```

Comment:
Why do we need this flag now?

Reply:
Do you mean `finishOnErr`? If so, this is because, previously an exception would be thrown if yielding next item raises. In `SafeAsyncIter`, `next` is not allowed to throw anything but `CancelledError` and the error is communicated by returning `failure`. This flag allows the user to choose what to do with that failure - should the failure be just yielded and allowing iterator to continue, or shall it stop (the default).